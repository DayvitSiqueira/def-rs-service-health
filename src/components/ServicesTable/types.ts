import { TableProps as FlowbiteTableProps } from 'flowbite-react'
import { ServiceDataParams } from '@/@types'

export interface HeaderProps {
  name: string
  isFixed?: boolean
}

export interface ServicesTableProps extends FlowbiteTableProps {
  header: HeaderProps[]
  tableData: ServiceDataParams[] | undefined
  isLoading: boolean
  isError: boolean
}
