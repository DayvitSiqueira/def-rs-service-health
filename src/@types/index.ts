export type StatusServiceParams = 'up' | 'error' | 'warning'

export type ServiceDataParams = {
  id: number
  service: string
  status: StatusServiceParams
  message: string
}
