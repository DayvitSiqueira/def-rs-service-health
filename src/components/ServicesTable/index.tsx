import ErrorComponent from '@/components/ErrorComponent'
import { ServicesTableProps } from './types'
import LoadingTable from './components/LoadingTable'
import { renderCurrentSelection } from './helpers/renderBadgeTable'

export default function ServicesTable({
  header,
  tableData,
  isLoading,
  isError,
}: ServicesTableProps) {
  let content

  if (isError) {
    content = <ErrorComponent />
  } else if (isLoading) {
    content = <LoadingTable />
  } else {
    content = (
      <table className="w-full text-sm text-left text-gray-500 dark:text-gray-400">
        <thead className="text-xs text-gray-700 uppercase dark:text-gray-400 text-center">
          <tr>
            {header.map((item, index, array) => (
              <th
                key={item.name}
                scope="col"
                className={`p-4 whitespace-nowrap bg-gray-100 dark:bg-gray-800 border-b-2
                  ${index === array.length - 1 && 'sticky right-0'}
                `}
              >
                {item.name}
              </th>
            ))}
          </tr>
        </thead>

        <tbody>
          {tableData?.map(item => (
            <tr
              key={item.id}
              className="border-b border-gray-200 dark:border-gray-700 text-center"
            >
              <td className="p-4">{item.service}</td>
              <td className="p-4">{renderCurrentSelection(item.status)}</td>
              <td className="p-4">{item.message}</td>
            </tr>
          ))}
        </tbody>
      </table>
    )
  }

  return (
    <div className="relative overflow-x-auto shadow-md sm:rounded-lg w-full">
      {content}
    </div>
  )
}
