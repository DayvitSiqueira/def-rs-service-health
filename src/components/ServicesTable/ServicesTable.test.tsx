import React from 'react'
import { render, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import { ServiceDataParams } from '@/@types'
import ServicesTable from './index'

const mockServicesHealthHeader = [
  { name: 'Serviço', isFixed: false },
  { name: 'Status', isFixed: false },
  { name: 'Observação', isFixed: false },
]

const mockServicesTableMock: ServiceDataParams[] = [
  {
    id: 1,
    service: 'SERPRO',
    status: 'up',
    message: '',
  },
  {
    id: 2,
    service: 'Conselho Nacional de Justiça',
    status: 'warning',
    message: 'Pode apresentar lentidão - Manutenção Preventiva até 15/06/2023',
  },
  {
    id: 3,
    service: 'Tribunal de Justiça',
    status: 'error',
    message: 'Sem privisão de normalização',
  },
  {
    id: 4,
    service: 'Gov.BR',
    status: 'up',
    message: 'Manutenção preventiva prevista para 17/07/2023',
  },
]

describe('ServicesTable', () => {
  it('renders the ServicesTable component correctly with data', () => {
    render(
      <ServicesTable
        header={mockServicesHealthHeader}
        tableData={mockServicesTableMock}
        isLoading={false}
        isError={false}
      />
    )

    const tableElement = screen.getByRole('table')
    const thElements = screen.getAllByRole('columnheader')
    const tdElements = screen.getAllByRole('cell')

    expect(tableElement).toBeInTheDocument()
    expect(thElements.length).toBe(3)
    expect(tdElements.length).toBe(12)
  })

  it('renders the ServicesTable component correctly when loading', () => {
    render(
      <ServicesTable
        header={mockServicesHealthHeader}
        tableData={[]}
        isLoading
        isError={false}
      />
    )

    const loadingTableElement = screen.getByRole('table')
    const loadingThElements = screen.getAllByRole('columnheader')
    const loadingTdElements = screen.getAllByRole('cell')

    expect(loadingTableElement).toBeInTheDocument()
    expect(loadingThElements.length).toBe(3)
    expect(loadingTdElements.length).toBe(3)
  })

  it('renders the ServicesTable component correctly when there is an error', () => {
    render(
      <ServicesTable
        header={mockServicesHealthHeader}
        tableData={[]}
        isLoading={false}
        isError
      />
    )

    const errorComponentElement = screen.getByTestId('error-component')

    expect(errorComponentElement).toBeInTheDocument()
  })

  it('matches the snapshot', () => {
    const { container } = render(
      <ServicesTable
        header={mockServicesHealthHeader}
        tableData={mockServicesTableMock}
        isLoading={false}
        isError={false}
      />
    )
    expect(container).toMatchSnapshot()
  })
})
