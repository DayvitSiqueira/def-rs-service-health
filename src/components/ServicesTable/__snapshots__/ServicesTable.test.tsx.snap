// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServicesTable matches the snapshot 1`] = `
<div>
  <div
    class="relative overflow-x-auto shadow-md sm:rounded-lg w-full"
  >
    <table
      class="w-full text-sm text-left text-gray-500 dark:text-gray-400"
    >
      <thead
        class="text-xs text-gray-700 uppercase dark:text-gray-400 text-center"
      >
        <tr>
          <th
            class="p-4 whitespace-nowrap bg-gray-100 dark:bg-gray-800 border-b-2
                  false
                "
            scope="col"
          >
            Serviço
          </th>
          <th
            class="p-4 whitespace-nowrap bg-gray-100 dark:bg-gray-800 border-b-2
                  false
                "
            scope="col"
          >
            Status
          </th>
          <th
            class="p-4 whitespace-nowrap bg-gray-100 dark:bg-gray-800 border-b-2
                  sticky right-0
                "
            scope="col"
          >
            Observação
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="border-b border-gray-200 dark:border-gray-700 text-center"
        >
          <td
            class="p-4"
          >
            SERPRO
          </td>
          <td
            class="p-4"
          >
            <span
              class="bg-green-100 text-green-800 text-xs font-medium mr-2 px-2.5 py-0.5 rounded dark:bg-green-900 dark:text-green-300"
            >
              Disponível
            </span>
          </td>
          <td
            class="p-4"
          />
        </tr>
        <tr
          class="border-b border-gray-200 dark:border-gray-700 text-center"
        >
          <td
            class="p-4"
          >
            Conselho Nacional de Justiça
          </td>
          <td
            class="p-4"
          >
            <span
              class="bg-yellow-100 text-yellow-800 text-xs font-medium mr-2 px-2.5 py-0.5 rounded dark:bg-yellow-900 dark:text-yellow-300"
            >
              Instável
            </span>
          </td>
          <td
            class="p-4"
          >
            Pode apresentar lentidão - Manutenção Preventiva até 15/06/2023
          </td>
        </tr>
        <tr
          class="border-b border-gray-200 dark:border-gray-700 text-center"
        >
          <td
            class="p-4"
          >
            Tribunal de Justiça
          </td>
          <td
            class="p-4"
          >
            <span
              class="bg-red-100 text-red-800 text-xs font-medium mr-2 px-2.5 py-0.5 rounded dark:bg-red-900 dark:text-red-300"
            >
              Indisponível
            </span>
          </td>
          <td
            class="p-4"
          >
            Sem privisão de normalização
          </td>
        </tr>
        <tr
          class="border-b border-gray-200 dark:border-gray-700 text-center"
        >
          <td
            class="p-4"
          >
            Gov.BR
          </td>
          <td
            class="p-4"
          >
            <span
              class="bg-green-100 text-green-800 text-xs font-medium mr-2 px-2.5 py-0.5 rounded dark:bg-green-900 dark:text-green-300"
            >
              Disponível
            </span>
          </td>
          <td
            class="p-4"
          >
            Manutenção preventiva prevista para 17/07/2023
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
